Concurrent Support

xsun2001 opened the pull request.

Commit:
Init CMake support

Commit:
add doxygen support to cmake

Commit:
Init thread support

Commit:
fix windows build failure
rename directory name
polish visual studio support

creatorlxd commented on Doxyfile:
```diff
@@ -17,6 +17,9 @@
 #---------------------------------------------------------------------------
 
+# Set input directories
+INPUT = /Source
```
add .

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+*/
+
+#ifndef SPACEGAMEENGINE_THREAD_H
+#define SPACEGAMEENGINE_THREAD_H
```
do not use ifndef
use #pragma once

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+{
+	using ThreadID = std::thread::id;
+
+	/*!\brief Class SpaceGameEngine::Thread is a RAII manager of a std::thread.
```
rename manager

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+
+		~Thread();
+
+		Thread &operator=( Thread &&other ) noexcept = delete;
```
other three functions
support move(2 function)

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+
+		Thread &operator=( Thread &&other ) noexcept = delete;
+
+		bool isJoinable() const noexcept;
```
rename

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+	{
+	public:
+		template<class Function, class... Args>
+		explicit Thread( Function &&f, Args &&... args ) : threadImpl( f, args... ) {}
```
add forward

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+
+		void Swap( Thread &other ) noexcept;
+
+		ThreadID GetThreadID() const noexcept;
```
use Id

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+		static ThreadID GetCurrentThreadID() noexcept;
+
+		template<class Rep, class Period>
+		static void Sleep( const std::chrono::duration<Rep, Period> &sleep_duration )
```
change std::chrono ...

creatorlxd commented on Source/Common/Public/concurrent/Thread.h:
```diff
@@ -0,0 +1,85 @@
+		}
+
+	private:
+		std::thread threadImpl;
```
rename

Commit:
bug fix and small tweaks
Doxyfile INPUT path reformat
use #pragma once instead of #ifndef checks
use std::forward() to pass arguments
add move support and some basic constructors
rename

Commit:
init mutex/lock/condition support
todo: document and unittest

Commit:
add Lock.h Lock.cpp to build path (sorry I forgot it)
fix move constructor and move assignment
add a simple test

xsun2001 commented:
@creatorlxd 有时间给我review一下。感觉问题很多。

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+		class ScopedLock;
+
+	private:
+		std::recursive_timed_mutex mutexImpl;
```
rename

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+	private:
+		Condition( std::unique_lock<std::recursive_timed_mutex>& lock );
+
+		std::unique_lock<std::recursive_timed_mutex>& lockImpl;
```
rename

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+		Condition( std::unique_lock<std::recursive_timed_mutex>& lock );
+
+		std::unique_lock<std::recursive_timed_mutex>& lockImpl;
+		std::condition_variable_any conditionImpl;
```
rename

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+
+		~Mutex() = default;
+
+		Mutex &operator=( const Mutex &&other ) = delete;
```
four function?only two here!

xsun2001 replied:
std::recursive_timed_mutex is NOT MoveConstructible and MoveAssignable. Compilers will not generate move constructors when copy constructors exist.

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+	class Condition
+	{
+	public:
+		Condition( const Condition & ) = delete;
```
four functions!!!

xsun2001 replied:
std::condition_variable_any is NOT MoveConstructible and MoveAssignable. Compilers will not generate move constructors when copy constructors exist.

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+
+		~Mutex() = default;
+
+		Mutex &operator=( const Mutex &&other ) = delete;
```
do not use const&&,just use &&

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+
+		void Wait();
+
+		void Wait( std::function<bool()> );
```
wait for using SpaceGameEngine::Function
or use template to support all callable object.

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+
+		void TryLock();
+
+		Condition newCondition();
```
rename

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+		void Unlock();
+
+	private:
+		std::unique_lock<std::recursive_timed_mutex> lockImpl;
```
rename

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,135 @@
+
+		~ScopedLock() = default;
+
+		ScopedLock &operator=( const ScopedLock &&other ) = delete;
```
four functions!!!!!!!!

xsun2001 replied:
std::scoped_lock is NOT MoveConstructible and MoveAssignable. Compilers will not generate move constructors when copy constructors exist.

creatorlxd reviewed (changes requested):
问题有点多，我不确定我是否都找到了。。。。

Commit:
rename member fields in Lock.h

Commit:
fix wrong copy constructor's signature.

Commit:
add missing newCondition() implement

Commit:
[ci skip] new tests
WARNING: ScopedLock is NOT functional

xsun2001 commented:
@creatorlxd 目前ScopedLock仍然无法编译。模版参数仍然有误。

我们仍然需要一个可以**生成含有特定个数类型的参数类型包**的模版元函数。比如:
```
gen_type_list<1>::type_list --> <some_type>
gen_type_list<2>::type_list --> <some_type, some_type>
gen_type_list<4>::type_list --> <some_type, some_type, some_type, some_type>
```

目前我可以找到的参考是clang里面的一个模版元函数：
```cpp
template <size_t _Ep, size_t _Sp = 0>
struct __make_tuple_indices
{
    static_assert(_Sp <= _Ep, "__make_tuple_indices input error");
    typedef __make_indices_imp<_Ep, _Sp> type;
};
```
它可以生成指定范围的tuple索引（正整数）数列。但是我并没有能力理解其内部实现：
```cpp
template <size_t _Ep, size_t _Sp>
using __make_indices_imp =
    typename __detail::__make<_Ep - _Sp>::type::template __to_tuple_indices<_Sp>;
```
```cpp
namespace __detail {

template<typename _Tp, size_t ..._Extra> struct __repeat;
template<typename _Tp, _Tp ..._Np, size_t ..._Extra> struct __repeat<__integer_sequence<_Tp, _Np...>, _Extra...> {
  typedef __integer_sequence<_Tp,
                           _Np...,
                           sizeof...(_Np) + _Np...,
                           2 * sizeof...(_Np) + _Np...,
                           3 * sizeof...(_Np) + _Np...,
                           4 * sizeof...(_Np) + _Np...,
                           5 * sizeof...(_Np) + _Np...,
                           6 * sizeof...(_Np) + _Np...,
                           7 * sizeof...(_Np) + _Np...,
                           _Extra...> type;
};

template<size_t _Np> struct __parity;
template<size_t _Np> struct __make : __parity<_Np % 8>::template __pmake<_Np> {};

template<> struct __make<0> { typedef __integer_sequence<size_t> type; };
template<> struct __make<1> { typedef __integer_sequence<size_t, 0> type; };
template<> struct __make<2> { typedef __integer_sequence<size_t, 0, 1> type; };
template<> struct __make<3> { typedef __integer_sequence<size_t, 0, 1, 2> type; };
template<> struct __make<4> { typedef __integer_sequence<size_t, 0, 1, 2, 3> type; };
template<> struct __make<5> { typedef __integer_sequence<size_t, 0, 1, 2, 3, 4> type; };
template<> struct __make<6> { typedef __integer_sequence<size_t, 0, 1, 2, 3, 4, 5> type; };
template<> struct __make<7> { typedef __integer_sequence<size_t, 0, 1, 2, 3, 4, 5, 6> type; };

template<> struct __parity<0> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type> {}; };
template<> struct __parity<1> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 1> {}; };
template<> struct __parity<2> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 2, _Np - 1> {}; };
template<> struct __parity<3> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 3, _Np - 2, _Np - 1> {}; };
template<> struct __parity<4> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 4, _Np - 3, _Np - 2, _Np - 1> {}; };
template<> struct __parity<5> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 5, _Np - 4, _Np - 3, _Np - 2, _Np - 1> {}; };
template<> struct __parity<6> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 6, _Np - 5, _Np - 4, _Np - 3, _Np - 2, _Np - 1> {}; };
template<> struct __parity<7> { template<size_t _Np> struct __pmake : __repeat<typename __make<_Np / 8>::type, _Np - 7, _Np - 6, _Np - 5, _Np - 4, _Np - 3, _Np - 2, _Np - 1> {}; };

} // namespace detail
```

> 以上为LLVM开源代码，在MIT 和 the University of Illinois Open Source Licenses许可下开源。

请协助我实现这一功能。

Commit:
change Condition api and finish tests

Commit:
rename ReentrantLock to RecursiveLock

Commit:
add doxygen document

Commit:
add todos

Commit:
add future.h

creatorlxd commented on Source/Common/Public/Concurrent/Lock.h:
```diff
@@ -0,0 +1,176 @@
+		/*!
+		 * \todo use SGE's function instead of std's
+		 */
+		void Wait( RecursiveLock &lock, std::function<bool()> );
```
you'd better give the second parameter a name which is same with its implement.

creatorlxd commented on Source/Common/Public/Concurrent/Thread.h:
```diff
@@ -0,0 +1,94 @@
+		Thread( Thread && ) noexcept;
+
+		template<class Function, class... Args>
+		explicit Thread( Function &&f, Args &&... args ) : m_ThreadImpl( std::forward<Function>( f ),
```
where the else?????

creatorlxd replied:
or format it correctly.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+	public:
+		enum class FutureState : int8_t
+		{
+			Working = 0, Succeed = 1, Failed = 2
```
please new line...

creatorlxd replied:
or format it correctly

xsun2001 replied:
？

creatorlxd replied:
just write like:
```
Working=0,
Succeed=1,
Failed=2
```

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+	{
+		Mutex m_Mutex;
+		ResultType m_Result;
+		std::atomic_int8_t m_State;
```
need we wrap atomic types?

xsun2001 replied:
I just **use** atomic int. It is essential in concurrent environment.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+		ResultType Get()
+		{
+			Wait();
+			SGE_CHECK( ExecutionError, GetState() == FutureState::Failed );
```
invalid usage.
please make this expr in the xxxerror::judge.
do not just repeat assert.

xsun2001 replied:
No.  it’s only one of ways to tigger the error.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+		template<class Rep, class Period>
+		ResultType Get( const std::chrono::duration<Rep, Period> &rel_time )
+		{
+			SGE_CHECK( TimeoutError, !Wait( rel_time ));
```
same to the previous.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+		ResultType Get( const std::chrono::duration<Rep, Period> &rel_time )
+		{
+			SGE_CHECK( TimeoutError, !Wait( rel_time ));
+			SGE_CHECK( ExecutionError, GetState() == FutureState::Failed );
```
same to the previous.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Execution Error" );
+
+		static bool Judge( const bool isFailed )
```
you can make the Judge into .cpp file.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,104 @@
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Timeout" );
+
+		static bool Judge( const bool isTimeout )
```
same to the previous.

Commit:
small tweaks

Commit:
doxygen for Futurn.h

xsun2001 commented:
完成文档和代码审查后就可以合并第一阶段的工作了

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,154 @@
+	 */
+	struct ExecutionError
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Execution Error" );
```
this is not a good practice.I suggest you that you can use ThrowError instead of this,or just use a correct error defination,never just return true.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,154 @@
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Timeout" );
+
+		static bool Judge();
```
see previous.

creatorlxd commented on Source/Common/Public/Future.h:
```diff
@@ -0,0 +1,154 @@
+			if (GetState() == FutureState::Failed)
+			{
+				SGE_CHECK( ExecutionError );
+			}
```
bad usage of SGE_CHECK.

creatorlxd reviewed (changes requested):
error handling functions have some problems.
waiting for your offical document in Document directory.

creatorlxd commented:
@xsun2001 你的错误处理还要改改
而且你忘了在头文件的命名空间里加doxygen的模块声明
你看我的代码学一下

Commit:
add doxygen module declaration

xsun2001 commented:
关于错误处理：
这些错误都是运行时**可以**出现的错误，程序应该对去进行相应的处理。目前的错误处理仍然无法捕获错误，而是直接使程序崩溃退出。目前我将不会对其进行更改，并等到下一阶段改进这方面的api，比如使用错误回调以绕过SGE的错误机制（如果Error.h不进行改变的话）。

xsun2001 commented:
文档工作可能会延后，但是没有文档不会对使用造成太大困难。
审核完毕后可以合并了。

Commit:
finish document of concurrent stage 1

xsun2001 commented:
@creatorlxd 该分支上的所有工作均已初步完成。请对所有代码及文档做最后的审查。主要检查错误拼写，编码规范，代码格式，以及文档中不通顺、有歧义或与代码不符的地方。也请考虑目前的API命名和设计问题，以尽快修改。

creatorlxd commented:
@xsun2001 我最近可能有点忙，我会在有空时跟你联系，讨论你代码中的问题（尤其是错误处理）。
我最近可能要旅游耽误六七天，而且Function没你想象的那么好写，但也能写（牵涉到元信息（初级的））。

creatorlxd commented on Source/Common/Private/Future.cpp:
```diff
@@ -0,0 +1,27 @@
+
+#include "Future.h"
+
+bool SpaceGameEngine::TimeoutError::Judge()
```
这样的写法是不正规的！！！

xsun2001 commented:
1.鉴于目前错误机制的功能缺失，并且你不同意在下一阶段进行进一步更改，我将会在该分支上弃用SGE的错误处理。
2.我并没有说Function难写。
3.请尽快开始元信息的编写。
4.请尽快开始构建系统的编写。下一工作将会涉及更多的第三方依赖。
5.在此分支合并后，我会立刻开始下一步工作，预期开启三个分支的同步工作。

creatorlxd commented:
@xsun2001 
1. 错误处理不是功能缺失，我觉得你误解了我的意思。你的错误处理的方法有问题，我可以在qq上和你语音谈这个事。
2. 我把“好”打成了“难”
3. 我在写元信息（简单的）
4. 这个得等基础库。。。

xsun2001 commented:
1.我如何处理抛出/产生的错误？
2.......
4.那我先用python脚本处理一下

Commit:
remove Future.h in current branch.

xsun2001 commented:
目前不适合加入`Future`及其相关特性
请审阅其余部分 @creatorlxd

creatorlxd commented:
@xsun2001 其余都差不多了，就剩错误处理的问题了。虽然我觉得这个文档写的有点水。。。

xsun2001 commented:
有关错误处理的我都删除了
你想怎么写这个文档啊......其实都是废话，全删了就行，都是稍微正经的程序员都了解的事情。

creatorlxd commented:
文档的问题我们还是找个时间在qq上谈吧。
错误处理还是必要的，你只是需要改个写法而已。日后会将其完善的。@xsun2001

Commit:
remove unnecessary files
